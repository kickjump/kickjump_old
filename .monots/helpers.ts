import path from 'node:path';
import url from 'node:url';

const __dirname = path.dirname(url.fileURLToPath(import.meta.url));

/**
 * Get a path relative to the base directory of this project. If called with no
 * arguments it will return the base directory.
 */
export function baseDir(...paths: string[]): string {
  return path.resolve(__dirname, '..', path.join(...paths));
}
