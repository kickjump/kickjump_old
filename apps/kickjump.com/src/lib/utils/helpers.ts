import is from '@sindresorhus/is';

import type { TupleRange } from '$types';

/**
 * Create a range from start to end.
 *
 * If only start is provided it creates an array of the size provided. if start
 * and end are provided it creates an array who's first position is start and
 * final position is end. i.e. `length = (end - start) + 1`.
 *
 * If you'd like to create a typed tuple of up to `40` items then pass in a
 * `[number]` tuple as the first argument.
 */
export function range<Size extends number>(size: [Size]): TupleRange<Size>;
export function range(size: number): number[];
export function range(start: number, end: number): number[];
export function range(start: number | [number], end?: number): number[] {
  const startValue = is.array(start) ? start[0] : start;

  if (!is.number(end)) {
    return Array.from(
      { length: Math.abs(startValue) },
      (_, index) => (startValue < 0 ? -1 : 1) * index,
    );
  }

  if (startValue <= end) {
    return Array.from({ length: end + 1 - startValue }, (_, index) => index + startValue);
  }

  return Array.from({ length: startValue + 1 - end }, (_, index) => -1 * index + startValue);
}
