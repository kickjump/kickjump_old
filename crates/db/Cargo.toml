[package]
name = "db"
version = "0.0.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
prisma-client-rust = { git = "https://github.com/Brendonovich/prisma-client-rust", tag = "0.4.0" }
serde = { version = "1.0.137", features = ["derive"] }
tokio = { version = "1.18.2", features = ["macros"] }
