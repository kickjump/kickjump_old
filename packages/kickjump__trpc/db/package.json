{
  "exports": {
    ".": {
      "browser": "./../dist/db.browser.esm.js",
      "import": "./../dist/db.esm.js",
      "types": "./../dist/db.d.ts"
    },
    "./index.js": {
      "browser": "./../dist/db.browser.esm.js",
      "import": "./../dist/db.esm.js",
      "types": "./../dist/db.d.ts"
    }
  },
  "module": "../dist/db.esm.js",
  "browser": "../dist/db.browser.esm.js",
  "types": "../dist/db.d.ts"
}
