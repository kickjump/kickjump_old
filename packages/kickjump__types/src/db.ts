import type { CreatedAt, UpdatedAt } from '@kickjump/edgedb/types';
import type {
  ConditionalKeys,
  ConditionalPick,
  PartialDeep,
  Primitive,
  SetOptional,
  Simplify,
} from 'type-fest';

export type ExtractFields<Type> = {
  [Key in keyof Type]?:
    | true
    | (Type[Key] extends ArrayUnion<infer T extends WithId> ? ExtractFields<T> : never);
};

export function createTypeSafeFields<Type extends WithId>() {
  return <Shape extends ExtractFields<Type>>(shape: Shape): Simplify<Shape> => {
    return shape;
  };
}

export type TypeFromFields<Type extends WithId, Fields extends ExtractFields<Type>> = Simplify<{
  [Key in keyof Fields]: Key extends keyof Type
    ? Fields[Key] extends true
      ? Type[Key] extends WithId
        ? WithId
        : Type[Key] extends WithId[]
        ? WithId[]
        : Type[Key] extends string
        ? `${Type[Key]}`
        : Type[Key]
      : Type[Key] extends WithId
      ? Fields[Key] extends ExtractFields<Type[Key]>
        ? TypeFromFields<Type[Key], Fields[Key]>
        : never
      : Type[Key] extends Array<infer T extends WithId>
      ? Fields[Key] extends ExtractFields<T>
        ? Array<TypeFromFields<T, Fields[Key]>>
        : never
      : never
    : never;
}>;

export { AccountProvider, Status, Visibility } from '@kickjump/edgedb/types';

/**
 * Matches any primitive, `Date`, or `RegExp` value.
 */
type BuiltIns = Primitive | Date | RegExp;
type Native = BuiltIns | object;

export type DeepOmit<Type, Omitted extends string> = Type extends BuiltIns
  ? Type
  : Type extends object
  ? Type extends Array<infer Item>
    ? Item[] extends Type
      ? Array<DeepOmit<Item, Omitted>>
      : DeepObjectOmit<Type, Omitted>
    : DeepObjectOmit<Type, Omitted>
  : Type;

type DeepObjectOmit<
  Type extends object,
  Omitted extends string,
  Keys extends Exclude<keyof Type, Omitted> = Exclude<keyof Type, Omitted>,
  PartialKeys extends Keys extends Keys
    ? undefined extends Type[Keys]
      ? Keys
      : never
    : never = Keys extends Keys ? (undefined extends Type[Keys] ? Keys : never) : never,
  RequiredKeys extends Exclude<Keys, PartialKeys> = Exclude<Keys, PartialKeys>,
> = {
  [Key in PartialKeys]?: DeepOmit<Type[Key], Omitted>;
} & {
  [Key in RequiredKeys]: DeepOmit<Type[Key], Omitted>;
};

type DeepOptionalKey<Type, Optional extends string> = Type extends BuiltIns
  ? Type
  : Type extends object
  ? Type extends Array<infer Item>
    ? Item[] extends Type
      ? Array<DeepOptionalKey<Item, Optional>>
      : DeepObjectOptionalKey<Type, Optional>
    : DeepObjectOptionalKey<Type, Optional>
  : Type;

type DeepObjectOptionalKey<
  Type,
  Optional extends string,
  Keys extends Optional extends keyof Type ? Optional : never = Optional extends keyof Type
    ? Optional
    : never,
  OtherKeys extends Exclude<keyof Type, Optional> = Exclude<keyof Type, Optional>,
  PartialKeys extends OtherKeys extends OtherKeys
    ? undefined extends Type[OtherKeys]
      ? OtherKeys
      : never
    : never = OtherKeys extends OtherKeys
    ? undefined extends Type[OtherKeys]
      ? OtherKeys
      : never
    : never,
  RequiredKeys extends Exclude<OtherKeys, PartialKeys> = Exclude<OtherKeys, PartialKeys>,
> = {
  [Key in Keys]?: DeepOptionalKey<Type[Key], Optional>;
} & {
  [Key in PartialKeys]?: DeepOptionalKey<Type[Key], Optional>;
} & {
  [Key in RequiredKeys]: DeepOptionalKey<Type[Key], Optional>;
};

export type UpdatedAtType<Options extends AvailableOptions = object> = Custom<
  DeepOmit<UpdatedAt, '__type__'>,
  Options
>;
export type CreatedAtType<Options extends AvailableOptions = object> = Custom<
  DeepOmit<CreatedAt, '__type__'>,
  Options
>;
export type CreateOmitKeys = 'id' | 'createdAt' | 'updatedAt' | 'actions';
export type UpdateOmitKeys = 'createdAt' | 'updatedAt' | 'actions';

export interface AvailableOptions {
  /** Deep omit keys */
  deepOmit?: string | undefined;
  /** Make every property optional in a deeply nested way. */
  deepPartial?: true | undefined;
  /** The keys to make optional for every nested object (can be used with id) */
  deepOptional?: string | undefined;
  /** Set the top level object to optional */
  partial?: true | undefined;
  /** Set the specified top level keys to optional */
  optional?: string | undefined;
  /** Omit top level keys */
  omit?: string | undefined;
  /** Pick top level keys */
  pick?: string | undefined;
  /** Conditionally omit properties which extend the provided */
  conditionalOmit?: Native | undefined;
  /** Conditionally pick properties which extend the provided type */
  conditionalPick?: Native | undefined;
  /** Remove all flattened object types */
  simplify?: boolean | undefined;
  /** Only hold onto nested database objects */
  complexify?: boolean | undefined;
  /** Replace the matching keys */
  replace?: object | undefined;
}

export type Custom<Type, Options extends AvailableOptions> = Options extends { deepPartial: true }
  ? Custom<PartialDeep<Type>, Omit<Options, 'deepPartial'>>
  : Options extends { simplify: true }
  ? Custom<ConditionalOmit<Type, ArrayUnion<WithId>>, Omit<Options, 'simplify'>>
  : Options extends { complexify: true }
  ? Custom<ConditionalPick<Type, ArrayUnion<WithId>>, Omit<Options, 'complexify'>>
  : Options extends { deepOmit: infer Keys extends string }
  ? Custom<DeepOmit<Type, Keys>, Omit<Options, 'deepOmit'>>
  : Options extends { deepOptional: infer Keys extends string }
  ? Custom<DeepOptionalKey<Type, Keys>, Omit<Options, 'deepOptional'>>
  : Options extends { partial: true }
  ? Custom<Partial<Type>, Omit<Options, 'partial'>>
  : Options extends { optional: infer Keys extends string }
  ? Custom<SetOptional<Type, Keys extends keyof Type ? Keys : never>, Omit<Options, 'optional'>>
  : Options extends { omit: infer Keys extends string }
  ? Custom<Omit<Type, Keys>, Omit<Options, 'omit'>>
  : Options extends { pick: infer Keys extends keyof Type }
  ? Custom<Pick<Type, Keys>, Omit<Options, 'pick'>>
  : Options extends { conditionalOmit: infer Shape }
  ? Custom<ConditionalOmit<Type, Shape>, Omit<Options, 'conditionalOmit'>>
  : Options extends { conditionalPick: infer Shape }
  ? Custom<ConditionalPick<Type, Shape>, Omit<Options, 'conditionalPick'>>
  : Options extends { replace: infer Replacement extends object }
  ? Custom<Replace<Type, Replacement>, Omit<Options, 'replace'>>
  : Simplify<Type>;

type Replace<Type, Replacement extends object> = Omit<Type, keyof Replacement> & Replacement;
type ConditionalOmit<Base, Condition> = Omit<Base, ConditionalKeys<Base, Condition>>;

export interface WithId {
  id: string;
}
export type ArrayUnion<Type> = Type[] | Type;
export type EnumUnion<Type extends string> = Type | `${Type}`;
