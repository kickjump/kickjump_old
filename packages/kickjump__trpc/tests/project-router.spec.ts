import { e, run, Visibility } from '@kickjump/db';
import { describe, expect, it } from 'vitest';

import { router } from '../src/index.js';
import { createContext } from './helpers.js';

describe('create', () => {
  it('requires authentication', async () => {
    const caller = router.createCaller(createContext());
    await expect(
      caller.project.create({ name: 'simple', description }),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      '"This procedure requires authentication to access"',
    );
  });

  it('can create a project', async () => {
    const { caller } = await setupAuthenticatedRouterCaller();
    const name = 'simple';

    const result = await caller.project.create({ name, description });

    const project = await getTestableProject(name);

    expect(result).toBeTypeOf('string');
    expect(project).toMatchSnapshot();
  });

  it('can create a project with tags', async () => {
    const { caller } = await setupAuthenticatedRouterCaller();
    const name = 'simple';
    await caller.project.create({ name, description, tags: ['welcome', 'to', 'the', 'jungle'] });

    const project = await getTestableProject(name);
    expect(project).toMatchSnapshot();
  });
});

describe('update', () => {
  it('can update the description', async () => {
    const { caller } = await setupAuthenticatedRouterCaller();
    const name = 'simple';
    const id = await caller.project.create({ name, description });

    await caller.project.update({
      id,
      description: 'What about another description with enough characters?',
    });

    await expect(getTestableProject(name)).resolves.toMatchSnapshot();
  });

  it('can update tags', async () => {
    const { caller } = await setupAuthenticatedRouterCaller();
    const name = 'simple';
    const id = await caller.project.create({ name, description, tags: ['xyz'] });

    await caller.project.update({
      id,
      tags: { add: ['abc', 'def', 'ghi', 'jkl'], remove: ['xyz'] },
    });

    await expect(getTestableProject(name)).resolves.toMatchSnapshot();

    await caller.project.update({ id, tags: { removeAll: true } });
    const result = await getTestableProject(name);
    expect(result?.tags).toHaveLength(0);
  });

  describe('permissions', () => {
    it('only lets users with required permissions update', async () => {
      const name = 'simple';
      const newDescription = 'An alternative title to test with.';
      const [nonMember, member, admin] = await run(
        e.set(
          e.insert(e.User, { username: 'john' }),
          e.insert(e.User, { username: 'wendy' }),
          e.insert(e.User, { username: 'sally' }),
        ),
      );
      const { caller } = await setupAuthenticatedRouterCaller();
      const id = await caller.project.create({ name, description });
      const noMembership = router.createCaller(createContext({ user: { id: nonMember.id } }));
      await expect(
        noMembership.project.update({
          id,
          description: newDescription,
          visibility: Visibility.$public,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();

      const memberQuery = e.insert(e.Membership, {
        actor: e.select(e.User, (user) => ({ filter: e.op(user.id, '=', e.uuid(member.id)) })),
        entity: e.select(e.Project, (project) => ({ filter: e.op(project.id, '=', e.uuid(id)) })),
        permissions: [Visibility.member],
      });

      await run(memberQuery);

      const memberCaller = router.createCaller(createContext({ user: { id: member.id } }));
      await expect(
        memberCaller.project.update({
          id,
          description: newDescription,
          visibility: Visibility.$public,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();

      const admin3Query = e.insert(e.Membership, {
        actor: e.select(e.User, (user) => ({ filter: e.op(user.id, '=', e.uuid(admin.id)) })),
        entity: e.select(e.Project, (project) => ({ filter: e.op(project.id, '=', e.uuid(id)) })),
        permissions: [Visibility.admin],
      });
      await run(admin3Query);

      const adminCaller = router.createCaller(createContext({ user: { id: admin.id } }));
      await expect(
        adminCaller.project.update({
          id,
          description: newDescription,
          visibility: Visibility.$public,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();

      await caller.project.update({ id, visibility: Visibility.member });
      await adminCaller.project.update({
        id,
        description: newDescription,
        visibility: Visibility.$public,
      });

      await expect(getTestableProject(name)).resolves.toMatchSnapshot();
    });
  });
});

const description = 'This needs to be at least 50 characters';

async function createUserWithAccount() {
  const username = 'test-user';
  const accessToken = 'this-is-an-access-token';
  const user = e.insert(e.User, { username });
  const account = e.insert(e.Account, {
    provider: 'github',
    providerAccountId: 'asdf',
    user,
    accessToken,
    login: 'test-user',
  });

  const [{ id }] = await run(e.set(user, account));

  return { username, accessToken, id };
}

async function setupAuthenticatedRouterCaller() {
  const { id } = await createUserWithAccount();
  const caller = router.createCaller(createContext({ user: { id } }));
  return { caller, id };
}

/** No ids or dates which change between test runs */
async function getTestableProject(name: string) {
  const projectQuery = e.select(e.Project, (project) => ({
    filter: e.op(project.name, '=', name),
    name: true,
    description: true,
    tags: { name: true },
    members: {
      permissions: true,
    },
    visibility: true,
  }));

  return await run(projectQuery);
}
