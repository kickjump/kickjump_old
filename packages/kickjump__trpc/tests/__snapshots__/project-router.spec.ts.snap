// Vitest Snapshot v1

exports[`create > can create a project 1`] = `
{
  "description": "This needs to be at least 50 characters",
  "members": [
    {
      "permissions": [
        "owner",
      ],
    },
  ],
  "name": "simple",
  "tags": [],
  "visibility": "owner",
}
`;

exports[`create > can create a project with tags 1`] = `
{
  "description": "This needs to be at least 50 characters",
  "members": [
    {
      "permissions": [
        "owner",
      ],
    },
  ],
  "name": "simple",
  "tags": [
    {
      "name": "welcome",
    },
    {
      "name": "to",
    },
    {
      "name": "the",
    },
    {
      "name": "jungle",
    },
  ],
  "visibility": "owner",
}
`;

exports[`update > can update tags 1`] = `
{
  "description": "This needs to be at least 50 characters",
  "members": [
    {
      "permissions": [
        "owner",
      ],
    },
  ],
  "name": "simple",
  "tags": [
    {
      "name": "abc",
    },
    {
      "name": "def",
    },
    {
      "name": "ghi",
    },
    {
      "name": "jkl",
    },
  ],
  "visibility": "owner",
}
`;

exports[`update > can update the description 1`] = `
{
  "description": "What about another description with enough characters?",
  "members": [
    {
      "permissions": [
        "owner",
      ],
    },
  ],
  "name": "simple",
  "tags": [],
  "visibility": "owner",
}
`;

exports[`update > permissions > only lets users with required permissions update 1`] = `"This user is not a member and has insufficient permissions to make requested changes."`;

exports[`update > permissions > only lets users with required permissions update 2`] = `
"Insufficient permissions to update field: 'visibility'.
Insufficient permissions to update field: 'description'."
`;

exports[`update > permissions > only lets users with required permissions update 3`] = `
"Insufficient permissions to update field: 'visibility'.
Insufficient permissions to update field: 'description'."
`;

exports[`update > permissions > only lets users with required permissions update 4`] = `
{
  "description": "An alternative title to test with.",
  "members": [
    {
      "permissions": [
        "owner",
      ],
    },
    {
      "permissions": [
        "member",
      ],
    },
    {
      "permissions": [
        "admin",
      ],
    },
  ],
  "name": "simple",
  "tags": [],
  "visibility": "public",
}
`;
